### DISCUSSION TITLE
[FR] Decmial Value Purchase and tracking

### DISCUSSION THREAD

**Discussion by huseyinkogo** [inventree/InvenTree#2554] [General]
Hey,

**Wondering if Inventree is able to do this?**

1. **Placing a purchase order of Milk around 1,242 Litres**
2. **Receiving the order, however the milk comes in enclosed cylinder tubs, Sometimes i receive the milk in tubs in set weight but most of the time it's different weight for each tub. I would like to track each tub, on which tubs were used in which production (build) or changed location.  Also all these tubs come with supplier barcodes.**

3. **How can I individually label them and track/scan?** 

> Using the current supplier’s barcode by scanning each barcode into a excel and decoding the barcode to its weight and importing the received values into inventree through Admin

> Using the current supplier’s barcode by scanning each barcode into an excel and decoding the barcode to its weight and importing  the weights like the above method and printing Inventree qr code for each tube.

4. **Are you able to bulk scan the items and change location with the app? (will purchase the app, when i learn this app in depth)**

Happy to donate for a feature like this.

Loving Inventree
Many Thanks

**Comment by matmair** [comment]
@huseyinkogo Hi there! I moved this to discussions as it is more of a general discussion and less an explicit, single feature request. I will try to address the questions you have following.

**Comment by matmair** [comment]
>     1. **Placing a purchase order of Milk around 1,242 Litres**

To be clear here: do you mean 1a) 1.242 one thousand two hundred forty two Litres or 1b) 1,242 one litre and a bit

>     2. **Receiving the order, however the milk comes in enclosed cylinder tubs, Sometimes i receive the milk in tubs in set weight but most of the time it's different weight for each tub. I would like to track each tub, on which tubs were used in which production (build) or changed location.  Also all these tubs come with supplier barcodes.**
> 

You can change the quantity of a stock item after its creation so I would set up a QA - station in general where everything received goes through and the weight (quantity) gets adjusted. If you need that information later you could also log it into a parameter and run analyses later on that info.

>     3. **How can I individually label them and track/scan?**

You can use the barcode on the can if they are unique, if there is additional information in there you can write a barcode plugin that transforms the data into the fields you want. 
If I find time I can assist you in writing that, should only be a few lines of code. Can you provide a few sample codes and what is contained?

>     4. **Are you able to bulk scan the items and change location with the app? (will purchase the app, when i learn this app in depth)**

Questions about the app are a thing for @SchrodingersGat in general.
Depending on the nature of your business and how many things you want to scan in it might be a good idea to buy a cheap Bluetooth scanner.

In general, a lot of this stuff depends on if you do this 3 times a month for 5 cans or as a regular business. A lot of processes should be adjusted if it is a business, there is a lot of QA etc. that can be included easily. If you want we could try to set up a short chat/call together to find out what might fit your process.

**Reply by huseyinkogo** [reply]
Thanks for the insightful reply @matmair , hope you're doing well?
> 
> 
> To be clear here: do you mean 1a) 1.242 one thousand two hundred forty two Litres or 1b) 1,242 one litre and a bit

**Receiving (1a 1.242 one thousand two hundred forty two Litres) basically receiving in total of 1,242 Litres of milk in a purchase order.**

> You can change the quantity of a stock item after its creation so I would set up a QA - station in general where everything received goes through and the weight (quantity) gets adjusted. If you need that information later you could also log it into a parameter and run analyses later on that info.

**I don't understand what you are trying to say :(**


> You can use the barcode on the can if they are unique, if there is additional information in there you can write a barcode plugin that transforms the data into the fields you want.
> If I find time I can assist you in writing that, should only be a few lines of code. Can you provide a few sample codes and what is contained?

**Yeah all my suppliers have their own unique barcodes on the tubes (i source the milk from 20+ suppliers, basically 20 different barcodes) all their barcode will specify the weight of their tub. eg of sample barcode;**

**BARCODE: 10012022(0238)3457(284)2345**
- **10012022 is the milking date**
- **0238 is the time**
- **3457 is the litre value - 34.57 litres**
- **284 is the batch number**
- **2345 product line**




> Depending on the nature of your business and how many things you want to scan in it might be a good idea to buy a cheap Bluetooth scanner.
> 
> In general, a lot of this stuff depends on if you do this 3 times a month for 5 cans or as a regular business. A lot of processes should be adjusted if it is a business, there is a lot of QA etc. that can be included easily. If you want we could try to set up a short chat/call together to find out what might fit your process.

**Generally our production is around 2 tonnes so around 111 tubes needs be tracked. from receival to cool storage to production/build. I was think of scanning all items received with a Bluetooth scanner which is connected to a pc. Hoping on use the app to through the mobile to scan and change stock location.**

**On the next comment i will write in detail of how my new business wants to use Inventree for stock management and stock/batch tracking.**

**Reply by matmair** [reply]
We do not use thousands symbols, so that might also be part of the problem.
The supplied barcode would be easily decoded via a [barcode plugin](https://inventree.readthedocs.io/en/latest/extend/plugins/barcode/). The feature was there for longer but this implementation is introduced with 0.6.0, which should be released shortly.
In this sice a barcodescanner will help you keep your sanity as picture based detection can take a few seconds per barcode.

**Comment by huseyinkogo** [comment]
Hey @matmair ,

How I would like to use Inventree for Milkshake production.

**APPROACH**:
To track build of a product from Raw Material to Finished Product

I will need the decimal values issue resolved in order use this method issue #2552

**PROBLEM**:
I receive some of my raw materials in different weights instead of set weights. So therefore I want to know how much is used in production) in exact weight:

**FIRST**
Purchase order has been placed for 3400.90 Litres of MIlk.

**SECOND**
Order arrives to the Factory, where the milk is stored in tubs, where all tubs have different weight like:

- 23.53 lites
- 21.83 lites
- 24.43 lites
- 20.23 lites
- and so on;
- 

As each tub as its own different value of weight, also it has the suppliers own barcode like

**BARCODE: 10012022(0238)3457(284)2345**

- 10012022 is the milking date
- 0238 is the time
- 3457 is the litre value of the tub- 34.57 litres
- 284 is the batch number
- 2345 product line

I would like to track each tub individually. I really care about the weight of the tub used from stock and raw material traceability so the purchase order is important too.

It will be great to use the supplier's own barcode to track, instead creating new barcodes for each unit. 

**THIRD** Production Day 
Let's say there is a stock of 3400.90 Litres of MIlk, where around 1,100 litres of milk is needed for production to make vanilla milkshake;
Staff member goes to the milk stock location, and scans the tubs needed for 1100 litres so around 110 tubs and changes the location to production.
When the stock changes location to production, then I would see how much milk will be actually left in the default location and used in production.

We can use a bluetooth scanners to receive stock and use stock for production, where the bluetooth scanner could be connected to raspberry pi zero w for the production usage and for receiving we could have another scanner connected to the pc 

And i will use the app for general inventory tasks.

**Reply by matmair** [reply]
@huseyinkogo this usage scenario should be possible. We will need to look at #2552 but no big blockers from what I can see.

Hosting InvenTree publicly can be dangerous (e.g. if someone reuses the password on the instance, no MFA is used, DDOS, etc.). If possible I really recommend putting everything in a closed network and using either a web proxy with DDOS in front or VPNs to secure it. If you use the feature set of InvenTree you will quickly expose a lot of sensitive business information and trends into it so keeping the deployment secure should be paramount.

**Comment by huseyinkogo** [comment]
@matmair 
Seems like the scenario for individual tracking won't be possible. I think i need keep it separate from Inventree, where i will need to create new barcode scanning script which will be linked to raspberry pi & wireless scanner.

However the the issue #2552 will be great if it was resolved.

**Reply by matmair** [reply]
@huseyinkogo I do not see why it should not work. If you need help feel free to ping me. My contact info is in my gh profile.

**Reply by huseyinkogo** [reply]
thanks @matmair in the upcoming weeks, when i have the free time, i will try to open up another discussion  and description the problem i want to resolve. (in detail)

**Comment by SchrodingersGat** [comment]
@huseyinkogo I think that your use case should be mostly supported already by InvenTree functionality, even if it is a bit "non obvious".

Each tub can be considered an individual "stock item", where the number of litres contained is represented by the "quantity"

The manufacturer barcode can be used to individually scan and track each tub (as long as the barcodes are unique!)

Stock items can be assigned custom barcodes, separate to the "InvenTree" barcodes.

For interpreting the barcodes and extracting stock quantity, etc, then you have a number of ways you can approach this.

### API

The [InvenTree API](https://inventree.readthedocs.io/en/latest/extend/api/) provides all the data you need. An external application can request data from the InvenTree server and be integrated with a barcode scanner, etc, however you like.

There is an existing [python library](https://inventree.readthedocs.io/en/latest/extend/python/) for API integration too.

### Plugin

Alternatively, you could write a [custom plugin](https://inventree.readthedocs.io/en/latest/extend/plugins/) which runs on the InvenTree server itself, to implement this custom functionality.

--------------

Either approach should be able to get you the custom functionality you need. If you have any suggestions for improvements on the core InvenTree functionality they'd be warmly received, however this use case certainly feels outside the standard scope!

**Reply by huseyinkogo** [reply]
@SchrodingersGat thanks for your informative reply mate. n the upcoming weeks, when i have the free time, i will try to open up another discussion and description the problem i want to resolve. (in detail)

**Reply by SchrodingersGat** [reply]
No worries. I'm sure we can come up with something.
